x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,327 @@
+		}
+
+		awsConfig.Region = regionName
+		svcSQS := sqs.New(awsConfig)
+		svcS3 := s3.New(awsConfig)
+
+		// RECEIVE
+		receiveMessageInput := &sqs.ReceiveMessageInput{
+			QueueUrl:              &queueURL,
+			MessageAttributeNames: []string{"All"},
+			MaxNumberOfMessages:   &MaxNumberOfMessage,
+			VisibilityTimeout:     awssdk.Int64(20), // 20 seconds
+			WaitTimeSeconds:       &WaitTimeSecond,
+		}
+
+		req := svcSQS.ReceiveMessageRequest(receiveMessageInput)
+		output, errR := req.Send()
+		if errR != nil {
+			return
+		}
+
+		if len(output.Messages) > 0 {
+			events, messagesReceiptHandles, err := p.receiveMessages(queueURL, output.Messages, svcS3, svcSQS)
+			if err != nil {
+				p.logger.Error(errors.Wrap(err, "receiveMessages failed"))
+			}
+
+			for _, event := range events {
+				d = &util.Data{Event: *event}
+				err = forwarder.Send(d)
+				if err != nil {
+					p.logger.Error(errors.Wrap(err, "forwarder send failed"))
+				}
+			}
+
+			// TODO: When log message collection takes longer than 30s(default filebeat freq?),
+			//  sqs messages got read twice or more because it didn't get deleted fast enough.
+			// delete message after events are sent
+			err = deleteMessages(queueURL, messagesReceiptHandles, svcSQS)
+			if err != nil {
+				p.logger.Error(errors.Wrap(err, "deleteMessages failed"))
+			}
+		}
+	}
+}
+
+// Stop stops the input and all its harvesters
+func (p *Input) Stop() {
+	p.registry.Stop()
+	p.outlet.Close()
+}
+
+// Wait stops the s3 input.
+func (p *Input) Wait() {
+	p.Stop()
+}
+
+func getRegionFromQueueURL(queueURL string) (string, error) {
+	// get region from queueURL
+	// Example: https://sqs.us-east-1.amazonaws.com/627959692251/test-s3-logs
+	queueURLSplit := strings.Split(queueURL, ".")
+	if queueURLSplit[0] == "https://sqs" && queueURLSplit[2] == "amazonaws" {
+		return queueURLSplit[1], nil
+	}
+	return "", errors.New("queueURL is not in format: https://sqs.{REGION_ENDPOINT}.amazonaws.com/{ACCOUNT_NUMBER}/{QUEUE_NAME}")
+}
+
+// launches goroutine per received message and wait for all message to be processed
+func (p *Input) receiveMessages(queueURL string, messages []sqs.Message, svcS3 s3iface.S3API, svcSQS *sqs.SQS) ([]*beat.Event, []string, error) {
+	var eventsTotal []*beat.Event
+	var messagesReceiptHandles []string
+	var wg sync.WaitGroup
+
+	// TODO: Check goroutine cleanup
+	numMessages := len(messages)
+	wg.Add(numMessages)
+	for i := range messages {
+		go func(m sqs.Message) {
+			// launch goroutine to handle each message
+			defer wg.Done()
+
+			s3Infos, err := handleMessage(m)
+			if err != nil {
+				p.logger.Error(err.Error())
+			}
+
+			if err != nil {
+				p.logger.Error(err.Error())
+			}
+
+			// read from s3
+			events, err := readS3Object(svcS3, s3Infos)
+			if err != nil {
+				p.logger.Error(err.Error())
+			}
+
+			eventsTotal = append(eventsTotal, events...)
+			messagesReceiptHandles = append(messagesReceiptHandles, *m.ReceiptHandle)
+		}(messages[i])
+	}
+
+	wg.Wait()
```

Comment:
`receiveMessages` is blocking until all messages are processed. This makes this work in batches of 10, but that doesn't mean that you are always fetching 10 objects at the same time. I wonder how big these S3 files get, or what's the expected concurrency here.

There is an opportunity here to separate concerns using channels & workers if needed

Reply:
So the batch size of `receiveMessages` is controlled by `maxNumberOfMessage` in `ReceiveMessage` API. Maximum is 10 which means SQS will never return more than 10 messages per api call but might return fewer than that. 
I think s3 files can be very large depending on what type of log files they are. I will think about using channels and workers here to make sure concurrency.

Reply:
Turned out `maxNumberOfMessage` the maximum number of messages to return. SQS never returns more messages than this value. BUT when I tested this, SQS usually return fewer in one API call. I did a little bit digging and I found this: Due to the distributed nature of the queue, a weighted random set of machines is sampled on a ReceiveMessage call. That means only the messages on the sampled machines are returned.